walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,69 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Arrays;
+import java.util.ArrayList;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    if (request.getAttendees().size() <= 0) {
+        return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+        return Arrays.asList();
+    }
+    // check every *duration* length block of the day
+    int minutes = 0;
+    int length = 0;
+    int duration = (int) request.getDuration();
+    int start = TimeRange.getTimeInMinutes(0, 0);
+    boolean available;
+    int time;
+    TimeRange availableMeetingTime = null;
+    ArrayList<TimeRange> availableMeetingTimes = new ArrayList<TimeRange>();
+    TimeRange potentialMeetingTime;
+    while (minutes < (24 * 60)) {
+        available = true; 
+        time = minutes + duration;
+        potentialMeetingTime = TimeRange.fromStartDuration(minutes, duration);
+        // if meeting overlaps with any event, check if any attendees are conflicted
+        for (Event event: events) {
+            if (potentialMeetingTime.overlaps(event.getWhen())) {
+                for (String attendee: request.getAttendees()) {
+                    if (event.getAttendees().contains(attendee)) {
+                        available = false;
+                        break;
+                    }
+                }
+            }
+        }
+        if (available) {
+            availableMeetingTimes.add(potentialMeetingTime);
+            minutes = time;
+        } else {
+            minutes++;
+        }
+    }
+
+    ArrayList<TimeRange> mergedMeetings = new ArrayList<TimeRange>();
+    if (availableMeetingTimes.size() <= 0) {
+        return mergedMeetings;
+    } else if (availableMeetingTimes.size() == 1) {
+        mergedMeetings.add(availableMeetingTimes.get(0));
+        return mergedMeetings;
+    }
+    int size = 1;
+    start = availableMeetingTimes.get(0).start();
+    for (int i = 0; i < availableMeetingTimes.size() - 1; i++) {
+        if (availableMeetingTimes.get(i).end() == availableMeetingTimes.get(i + 1).start()) {
+            if (i == availableMeetingTimes.size() - 2) {
```

Comment:
I do. I don't have a conditional for when i!=size-2 because the behavior doesn't change in that instance
What I'm doing is iterating through all the meeting times that are available and I'm waiting until I reach a "gap" in time (which is why I keep up with the start variable and then add that timerange to merged meetings arraylist).

size-2 is an edge case because I'm at the last index of the unmerged times that I can safely traverse without going out of bounds (I compare i to i+1, so I need to stop at size-2). at size-2 I just directly compare since I know I'm at the end. But to directly answer the question, if I'm not at size-2 and there's no timegap, I can just continue to the next iteration